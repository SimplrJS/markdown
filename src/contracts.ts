export interface EmphasisOptions {
    useUnderscores?: boolean;
}

export interface CodeOptions {
    lang?: string;
}

export interface UnorderedListOptions {
    symbol?: UnorderedListSymbols;
}

export type UnorderedListSymbols = "*" | "+" | "-";

export interface TableOptions {
    removeColumnIfEmpty?: boolean;
    removeRowIfEmpty?: boolean;
}

export interface TableHeader {
    text: string;
    align: TableAlign;
}

export type TableAlign = "left" | "center" | "right" | "none";

export interface MarkdownList extends Array<string | MarkdownList> { }

export type HorizontalRuleSymbol = "-" | "*" | "_";

export interface BlockquoteOptions {
    escape: boolean;
    escapeCharacter: string;
}
