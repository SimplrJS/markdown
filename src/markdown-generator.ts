import * as S from "string";

import {
    CodeOptions,
    EmphasisOptions,
    TableHeader,
    TableOptions,
    UnorderedListOptions,
    UnorderedListSymbols,
    MarkdownList,
    HorizontalRuleSymbol
} from "./contracts";

import { ListGenerator } from "./generators/list-generator";
import { TableGenerator } from "./generators/table-generator";

export namespace MarkdownGenerator {
    export function Header(text: string, headerLevel: number, closing: boolean = false): string {
        if (!isFinite(headerLevel)) {
            throw Error("Heading number must be finite!");
        }
        if (headerLevel < 1 || headerLevel > 6) {
            throw Error("Heading number must be more than 0 and less than 7.");
        }

        const headers = S("#").repeat(headerLevel).s;
        const sanitizedText = S(text).trim().s;

        if (closing) {
            return `${headers} ${sanitizedText} ${headers}`;
        } else {
            return `${headers} ${sanitizedText}`;
        }
    }

    export function UnderlineHeader(text: string, headerLevel: number): string[] {
        if (!isFinite(headerLevel)) {
            throw Error("Heading number must be finite!");
        }
        if (headerLevel < 1 || headerLevel > 2) {
            throw Error("Heading number must be more than 0 and less than 2.");
        }

        const sanitizedText = S(text).trim().s;
        const header = headerLevel === 1 ? S("=").repeat(sanitizedText.length).s : S("-").repeat(sanitizedText.length).s;

        return [
            sanitizedText,
            header
        ];
    }

    export function Blockquote(text: string | string[]): string[] {
        let textLines: string[] = [];

        if (typeof text === "string") {
            textLines = S(text).trim().lines();
        } else {
            textLines = text;
        }

        const lines: string[] = [];

        textLines.forEach(x => {
            if (x === "") {
                lines.push(">");
                return;
            }
            /**
             * Trimming only right.
             * Because other markdown elements could have on the left side of the text spacing,
             * for example: Lists.
             */
            const sanitizedText = S(x).trimRight().s;
            lines.push(`> ${sanitizedText}`);
        });

        return lines;
    }

    export function Link(text: string, definitionName: string, definition: boolean): string;
    export function Link(text: string, url: string, linkTitle ?: string): string;
    export function Link(text: string, target: string, arg ?: string | boolean): string {
        const sanitizedText = S(text).trim().s;

        // Link to a link definition
        if (typeof arg === "boolean") {
            return `[${sanitizedText}][${target}]`;
        }

        // Normal link
        let targetText = target;

        if (arg != null) {
            targetText += ` "${arg}"`;
        }

        return `[${sanitizedText}](${targetText})`;
    }

    export function LinkDefinition(text: string, url: string, linkTitle ?: string): string {
        const sanitizedText = S(text).trim().s;
        let linkTitleText = "";

        if (linkTitle != null) {
            linkTitleText = `"${linkTitle}"`;
        }

        return S(`[${sanitizedText}]: ${url} ${linkTitleText}`).trim().s;
    }

    export function Image(altText: string, url: string, title ?: string): string {
        return "!" + Link(altText, url, title);
    }

    export function UnorderedList(list: MarkdownList, options ?: UnorderedListOptions): string[] {
        const allowedSymbols: UnorderedListSymbols[] = [
            "*",
            "+",
            "-"
        ];
        let symbol: UnorderedListSymbols = "*";

        if (options != null && options.symbol != null) {
            if (allowedSymbols.indexOf(options.symbol) === -1) {
                throw Error(`Unordered list: Symbol ${options.symbol} is not allowed. Please use: ${allowedSymbols.join(", ")} instead.`);
            }

            symbol = options.symbol;
        }

        return ListGenerator.RenderList(list, 0, false, symbol);
    }

    export function OrderedList(list: MarkdownList): string[] {
        return ListGenerator.RenderList(list, 0, true);
    }

    export function HorizontalRule(symbol ?: HorizontalRuleSymbol, length ?: number): string {
        const allowedSymbols: HorizontalRuleSymbol[] = [
            "*",
            "-",
            "_"
        ];
        const defaultSymbol: HorizontalRuleSymbol = "-";
        const defaultLength = 3;

        if (symbol != null && allowedSymbols.indexOf(symbol) === -1) {
            throw Error(`Horizontal Rule: Symbol ${symbol} is not allowed. Please use: ${allowedSymbols.join(", ")} instead.`);
        }

        if (length != null && length < 3) {
            throw Error(`Horizontal Rule: Minimum length is ${defaultLength}`);
        }

        const currentSymbol = symbol || defaultSymbol;
        const currentLength = length || defaultLength;

        return S(currentSymbol).pad(currentLength, currentSymbol).s;
    }

    export function Italic(text: string, options ?: EmphasisOptions): string {
        const sanitizedText = S(text).trim().s;

        if (options != null && options.useUnderscores) {
            return `_${sanitizedText}_`;
        }

        return `*${sanitizedText}*`;
    }

    export function Bold(text: string, options ?: EmphasisOptions): string {
        const sanitizedText = S(text).trim().s;

        if (options != null && options.useUnderscores) {
            return `__${sanitizedText}__`;
        }

        return `**${sanitizedText}**`;
    }

    export function StrikeThrough(text: string): string {
        const sanitizedText = S(text).trim().s;

        return `~~${sanitizedText}~~`;
    }

    /**
     * Github flavored markdown
     * @see https://help.github.com/articles/basic-writing-and-formatting-syntax/#quoting-code
     */
    export function InlineCode(text: string): string {
        const sanitizedText = S(text).trim().s;

        return `\`${sanitizedText}\``;
    }

    /**
     * Github flavored markdown
     * @see https://help.github.com/articles/basic-writing-and-formatting-syntax/#quoting-code
     */
    export function Code(text: string | string[], options ?: CodeOptions): string[] {
        let sanitizedText: string[] = [];

        if (typeof text === "string") {
            sanitizedText = S(text).trim().lines();
        } else {
            sanitizedText = text;
        }

        const codeBlockTag = "```";
        let header = codeBlockTag;

        if (options != null && options.lang != null) {
            header += options.lang;
        }

        return [
            header,
            ...sanitizedText,
            codeBlockTag
        ];
    }

    export function Table(headers: string[] | TableHeader[], content: string[][], options ?: TableOptions): string[] {
        return TableGenerator.RenderTable(headers, content, options);
    }

    /**
     * Escapes string.
     * @see https://daringfireball.net/projects/markdown/syntax#backslash
     */
    export function escapeString(text: string): string {
        return text.replace(/[\\\`\*\_\{\}\[\]\(\)\#\+\-\.\!]/g, "\\$&");
    }
}
