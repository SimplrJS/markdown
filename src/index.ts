import * as Contracts from "./contracts";

export * from "./markdown-generator";
export * from "./markdown-builder";
export {
    Contracts
};
