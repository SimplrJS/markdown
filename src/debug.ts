import { MarkdownGenerator } from "./markdown-generator";

function Main(): void {
    const headers: string[] = ["Property", "Value"];
    const rows = [
        ["Name"],
        ["Age"]
    ];

    const result = MarkdownGenerator.Table(headers, rows, { removeColumnIfEmpty: true });
    // tslint:disable-next-line:no-debugger
    debugger;
}

Main();
