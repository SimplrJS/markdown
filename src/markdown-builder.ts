import { MarkdownGenerator } from "./markdown-generator";

import {
    CodeOptions,
    EmphasisOptions,
    TableHeader,
    TableOptions,
    UnorderedListOptions,
    UnorderedListSymbols,
    MarkdownList,
    HorizontalRuleSymbol
} from "./contracts";

export type TextHandler = (md: typeof MarkdownGenerator) => string | string[];
export type TextGenerator = string | string[] | TextHandler;

export class MarkdownBuilder {
    constructor(output?: string[] | MarkdownBuilder) {
        this.output = this.resolveInput(output || []);
    }

    private output: string[];

    public Header(text: string, headerLevel: number, closing: boolean = false): this {
        this.output.push(MarkdownGenerator.Header(text, headerLevel, closing));

        return this;
    }

    public UnderlineHeader(text: string, headerLevel: number): this {
        this.output = this.output.concat(MarkdownGenerator.UnderlineHeader(text, headerLevel));

        return this;
    }

    public Blockquote(text: string | string[] | MarkdownBuilder): this {
        this.output = this.output.concat(this.resolveInput(text));

        return this;
    }

    public Link(text: string, definitionName: string, definition: boolean): this;
    public Link(text: string, url: string, linkTitle?: string): this;
    public Link(text: string, target: string, arg?: string | boolean): this {
        // WTF?
        if (typeof arg === "boolean") {
            this.output.push(MarkdownGenerator.Link(text, target, arg));
        } else {
            this.output.push(MarkdownGenerator.Link(text, target, arg));
        }

        return this;
    }

    public LinkDefinition(text: string, url: string, linkTitle?: string): this {
        this.output.push(MarkdownGenerator.LinkDefinition(text, url, linkTitle));

        return this;
    }

    public Image(altText: string, url: string, title?: string): this {
        this.output.push(MarkdownGenerator.Image(altText, url, title));

        return this;
    }

    public UnorderedList(list: MarkdownList, options?: UnorderedListOptions): this {
        this.output = this.output.concat(MarkdownGenerator.UnorderedList(list, options));

        return this;
    }

    public OrderedList(list: MarkdownList): this {
        this.output = this.output.concat(MarkdownGenerator.OrderedList(list));

        return this;
    }

    public HorizontalRule(symbol?: HorizontalRuleSymbol, length?: number): this {
        this.output.push(MarkdownGenerator.HorizontalRule(symbol, length));

        return this;
    }

    public Italic(text: string, options?: EmphasisOptions): this {
        this.output.push(MarkdownGenerator.Italic(text, options));

        return this;
    }

    public Bold(text: string, options?: EmphasisOptions): this {
        this.output.push(MarkdownGenerator.Bold(text, options));

        return this;
    }

    public StrikeThrough(text: string): this {
        this.output.push(MarkdownGenerator.StrikeThrough(text));

        return this;
    }

    /**
     * Github flavored markdown
     * @see https://help.github.com/articles/basic-writing-and-formatting-syntax/#quoting-code
     */
    public InlineCode(text: string): this {
        this.output.push(MarkdownGenerator.InlineCode(text));

        return this;
    }

    /**
     * Github flavored markdown
     * @see https://help.github.com/articles/basic-writing-and-formatting-syntax/#quoting-code
     */
    public Code(text: string | string[], options?: CodeOptions): this {
        this.output = this.output.concat(MarkdownGenerator.Code(text, options));

        return this;
    }

    public Table(headers: Array<string | TableHeader>, content: string[][], options?: TableOptions): this {
        this.output = this.output.concat(MarkdownGenerator.Table(headers, content, options));

        return this;
    }

    public Text(text: TextGenerator): this {
        this.output = this.output.concat(this.resolveInput(text));

        return this;
    }

    public NewLine(): this {
        this.output.push("");

        return this;
    }

    private resolveInput(input: string | string[] | TextHandler | MarkdownBuilder): string[] {
        if (typeof input === "function") {
            input = input(MarkdownGenerator);
        }

        if (Array.isArray(input)) {
            return input;
        } else if (input instanceof MarkdownBuilder) {
            return input.GetOutput();
        } else {
            return [input];
        }
    }

    public GetOutput(): string[] {
        return this.output;
    }
}

new MarkdownBuilder()
    .Text(md => md.Bold("Hello"));
