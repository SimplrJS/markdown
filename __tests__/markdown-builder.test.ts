import { MarkdownBuilder } from "../src/markdown-builder";

describe("MarkdownBuilder", () => {
    it("Simple builder with GetOutput", () => {
        const builder = new MarkdownBuilder()
            .Header("Builder simple example", 1)
            .EmptyLine()
            .Text("It's a simple markdown builder example.");

        expect(builder.GetOutput()).toMatchSnapshot();
    });

    it("Simple builder with Build", () => {
        const builder = new MarkdownBuilder()
            .Header("Builder simple example", 1)
            .EmptyLine()
            .Text("It's a simple markdown builder example.");

        expect(builder.Build()).toMatchSnapshot();
        expect(builder.Build("\n")).toMatchSnapshot();
    });

    it("All used methods", () => {
        const builder = new MarkdownBuilder()
            .UnderlineHeader("Foo", 1)
            .EmptyLine()
            .Header("Bar", 2, true)
            .Blockquote("blockquote")
            .EmptyLine()
            .Link("Link Text", "reference-text", true)
            .EmptyLine()
            .Link("Link Url", "https://domain.domain", "qwd")
            .EmptyLine()
            .LinkDefinition("reference", "https://domain.domain")
            .EmptyLine()
            .Image("Alternative text", "https://domain.domain")
            .EmptyLine()
            .UnorderedList(["One", "Two"])
            .EmptyLine()
            .OrderedList(["One", "Two"])
            .EmptyLine()
            .HorizontalRule()
            .EmptyLine()
            .Italic("Italic")
            .EmptyLine()
            .Bold("Bold")
            .EmptyLine()
            .StrikeThrough("Strike Through")
            .EmptyLine()
            .InlineCode("Inline code")
            .EmptyLine()
            .Code(`const a = "Hello World!";`, { lang: "typescript" })
            .EmptyLine()
            .Table(["A", "B"], [["a1", "b1"]])
            .EmptyLine()
            .Text("Hello World!")
            .Text("It's a simple markdown builder example.");

        expect(builder.GetOutput()).toMatchSnapshot();
    });

    it("Append another Builder while constructing a new one", () => {
        const oldBuilder = new MarkdownBuilder()
            .UnderlineHeader("Hello Header", 1);

        const newBuilder = new MarkdownBuilder(oldBuilder)
            .EmptyLine()
            .Text(md => `Hello ${md.Bold("World")}!`);

        expect(newBuilder).toMatchSnapshot();
    });
});
