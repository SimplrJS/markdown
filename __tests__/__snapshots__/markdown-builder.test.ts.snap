// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarkdownBuilder All used methods 1`] = `
Array [
  "Foo",
  "===",
  "",
  "## Bar ##",
  "blockquote",
  "",
  "[Link Text][reference-text]",
  "",
  "[Link Url](https://domain.domain \\"qwd\\")",
  "",
  "[reference]: https://domain.domain",
  "",
  "![Alternative text](https://domain.domain)",
  "",
  "* One",
  "* Two",
  "",
  "1. One",
  "2. Two",
  "",
  "---",
  "",
  "*Italic*",
  "",
  "**Bold**",
  "",
  "~~Strike Through~~",
  "",
  "\`Inline code\`",
  "",
  "\`\`\`typescript",
  "const a = \\"Hello World!\\";",
  "\`\`\`",
  "",
  "| A  | B  |",
  "| -- | -- |",
  "| a1 | b1 |",
  "",
  "Hello World!",
  "It's a simple markdown builder example.",
]
`;

exports[`MarkdownBuilder Append another Builder while constructing a new one 1`] = `
MarkdownBuilder {
  "output": Array [
    "Hello Header",
    "============",
    "",
    "Hello **World**!",
  ],
}
`;

exports[`MarkdownBuilder Simple builder 1`] = `
Array [
  "# Builder simple example",
  "",
  "It's a simple markdown builder example.",
]
`;
